File: reference/flutter-src/lib/data/classifier/mock_species_classifier.dart
import 'dart:math';

import '../../domain/classifier/species_classifier.dart';
import '../../domain/models/classification_result.dart';
import '../../domain/repositories/species_repository.dart';

/// ┌──────────────────────────────────────────────────────────────────────┐
/// │  STUB — MockSpeciesClassifier                                         │
/// │                                                                        │
/// │  Stands in for the real wingbeat model until the TFLite CNN lands.     │
/// │  It ignores the audio content and returns a plausible random result   │
/// │  (primary + runner-up + a wingbeat frequency near the species range),  │
/// │  after a short delay that mimics on-device inference.                  │
/// │                                                                        │
/// │  Replace with TfliteSpeciesClassifier — see                            │
/// │  data/classifier/tflite_species_classifier.dart. No caller changes.    │
/// └──────────────────────────────────────────────────────────────────────┘
class MockSpeciesClassifier implements SpeciesClassifier {
  MockSpeciesClassifier(this._species, {Random? random})
      : _random = random ?? Random();

  final SpeciesRepository _species;
  final Random _random;

  @override
  Future<void> load() async {
    // Nothing to load for the mock. The real model would read weights + labels.
  }

  @override
  Future<ClassificationResult> classify(AudioSample sample) async {
    // Simulate inference latency (~1.7s, matching the design's analyze state).
    await Future<void>.delayed(const Duration(milliseconds: 1700));

    final ids = _species.classifiableIds;
    final primaryId = ids[_random.nextInt(ids.length)];
    var runnerId = ids[_random.nextInt(ids.length)];
    while (runnerId == primaryId) {
      runnerId = ids[_random.nextInt(ids.length)];
    }

    final primary = _species.byId(primaryId)!;
    final runner = _species.byId(runnerId)!;

    final confidence = 78 + _random.nextInt(17); // 78–94
    final runnerConfidence =
        max(3, 100 - confidence - _random.nextInt(6));

    return ClassificationResult(
      primary: primary,
      runner: runner,
      confidence: confidence,
      runnerConfidence: runnerConfidence,
      wingbeatHz: _freqFor(primaryId),
    );
  }

  /// A believable measured frequency inside each species' known range.
  int _freqFor(String id) {
    switch (id) {
      case 'aedes':
        return 450 + _random.nextInt(251); // 450–700
      case 'culex':
        return 300 + _random.nextInt(151); // 300–450
      case 'anopheles':
        return 400 + _random.nextInt(201); // 400–600
      default:
        return 400 + _random.nextInt(300);
    }
  }

  @override
  Future<void> dispose() async {}
}

File: app/src/main/kotlin/com/mozzid/data/classifier/MockSpeciesClassifier.kt
package com.mozzid.data.classifier

import com.mozzid.domain.classifier.AudioSample
import com.mozzid.domain.classifier.SpeciesClassifier
import com.mozzid.domain.model.ClassificationResult
import com.mozzid.domain.repository.SpeciesRepository
import kotlinx.coroutines.delay
import kotlin.math.max
import kotlin.random.Random

/**
 * ┌────────────────────────────────────────────────────────────────────────┐
 * │  STUB — MockSpeciesClassifier                                           │
 * │                                                                          │
 * │  Stands in for the real wingbeat model until the TFLite CNN lands. It    │
 * │  ignores the audio content and returns a plausible random result         │
 * │  (primary + runner-up + a wingbeat frequency near the species range),     │
 * │  after a short delay that mimics on-device inference.                     │
 * │                                                                          │
 * │  Replace with TfliteSpeciesClassifier — no caller changes.               │
 * └────────────────────────────────────────────────────────────────────────┘
 */
class MockSpeciesClassifier(
    private val species: SpeciesRepository,
    private val random: Random = Random.Default,
) : SpeciesClassifier {

    override suspend fun load() {
        // Nothing to load for the mock. The real model would read weights + labels.
    }

    override suspend fun classify(sample: AudioSample): ClassificationResult {
        // Simulate inference latency (~1.7s, matching the design's analyze state).
        delay(1700)

        val ids = species.classifiableIds
        val primaryId = ids[random.nextInt(ids.size)]
        var runnerId = ids[random.nextInt(ids.size)]
        while (runnerId == primaryId) {
            runnerId = ids[random.nextInt(ids.size)]
        }

        val primary = species.byId(primaryId)!!
        val runner = species.byId(runnerId)!!

        val confidence = 78 + random.nextInt(17) // 78..94
        val runnerConfidence = max(3, 100 - confidence - random.nextInt(6))

        return ClassificationResult(
            primary = primary,
            runner = runner,
            confidence = confidence,
            runnerConfidence = runnerConfidence,
            wingbeatHz = freqFor(primaryId),
        )
    }

    /** A believable measured frequency inside each species' known range. */
    private fun freqFor(id: String): Int = when (id) {
        "aedes_aegypti", "aedes" -> 450 + random.nextInt(251)          // 450..700 Hz
        "aedes_albopictus" -> 450 + random.nextInt(201)                // 450..650 Hz
        "anopheles_gambiae", "anopheles" -> 400 + random.nextInt(201)  // 400..600 Hz
        "anopheles_albimanus" -> 400 + random.nextInt(181)             // 400..580 Hz
        "culex_pipiens" -> 300 + random.nextInt(151)                   // 300..450 Hz
        "culex_quinquefasciatus", "culex" -> 300 + random.nextInt(181) // 300..480 Hz
        else -> 300 + random.nextInt(400)
    }

    override suspend fun dispose() {}
}
